test/basics.jl

```diff
@@ -0,0 +1,60 @@
+test_string_length = 100
+
+#string_types = [ASCIIStr, LatinStr, UTF8Str, UCS2Str, UTF16Str, UTF32Str, UniStr]
+#string_types = [ASCIIStr, LatinStr, UTF8Str, UCS2Str, UTF16Str, UTF32Str]
+string_types = [ASCIIStr, LatinStr]
+
+##  create type specific test strings
+test_strings_base = Dict{String, Any}()
+for T in string_types
+    tmax = UInt32(typemax(Strs.codepoint_type(T)))
```

Comment:
if tmax > 0xd7ff, you must subtract from 0x800 from tmax, and then add back 0x800 if the returned random value is > 0xd7ff, to handle skipping over the surrogates.